File: Solutions/Python/design_add_and_search_words_data_structure.py
"""
    [MEDIUM]
    211. Design Add and Search Words Data Structure

    Concepts:
    - design
    - trie
    - depth-first search

    Stats:
        Runtime | 806 ms   [Beats 93.92%]
        Memory  | 52.62 MB  [Beats 84.66%]
"""

class WordDictionary:

    def __init__(self):
        self.nodes = {}

    def addWord(self, word: str) -> None:
        curr = self.nodes

        for c in word:
            if c not in curr:
                curr[c] = {}
            curr = curr[c]

        curr[None] = None

    def search(self, target: str) -> bool:
        
        def dfs(curr: dict[str], idx: int) -> bool:
            if not curr:
                return False

            # if entire target searched...
            if idx == len(target):
                return None in curr

            # if leaf reached before entire target searched...
            if target[idx] in curr:
                return dfs(curr[target[idx]], idx + 1)

            if target[idx] == ".":
                for c in curr:
                    if curr != None and dfs(curr[c], idx + 1):
                        return True

            return False

        return dfs(self.nodes, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: Solutions/Cpp/design_add_and_search_words_data_structure.cpp
/*
    [MEDIUM]
    211. Design Add and Search Words Data Structure

    Concepts:
    - design
    - trie
    - depth-first search

    Stats:
        Runtime | 778 ms    [Beats 13.68%]
        Memory  | 549.47 MB [Beats 72.89%]
*/

class TrieNode {
public:
    std::unordered_map<char, TrieNode*> nodes;
    bool is_leaf;

    TrieNode() {
        nodes = {};
        is_leaf = false;
    }
};

class WordDictionary {
public:
    TrieNode* root;

    WordDictionary() {
        root = new TrieNode();
    }
    
    void addWord(string word) {
        TrieNode* curr = this->root;

        for (const char c : word) {
            if (!curr->nodes.contains(c))
                curr->nodes[c] = new TrieNode();
            curr = curr->nodes[c];
        }
        curr->is_leaf = true;
    }
    
    bool search(string target) {
        std::function<bool(TrieNode*, int)> dfs;

        dfs = [&](TrieNode* curr, int idx){
            if (!curr)
                return false;

            // if entire target searched...
            if (idx == target.size())
                return curr->is_leaf;

            // if leaf reached before entire target searched...
            if (curr->nodes.contains(target[idx]))
                return dfs(curr->nodes[target[idx]], idx + 1);

            if (target[idx] == '.') {
                for (const auto& [c, node] : curr->nodes) {
                    if (curr && dfs(node, idx + 1))
                        return true;
                }
            }

            return false;
        };

        return dfs(this->root, 0);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */